phosphobot/am/act.py

```diff
@@ -1,40 +1,387 @@
+            raise ValueError(f"State key {state_key} not found in features")
+        if "state" not in state_key.lower():
+            raise ValueError(f"State key {state_key} must contain 'state'")
+        if features[state_key].type != "STATE":
+            raise ValueError(f"State key {state_key} must map to a STATE feature")
+
+        # Validate video_keys
+        for key in video_keys:
+            if key not in features:
+                raise ValueError(f"Image key {key} not found in features")
+            if "image" not in key.lower():
+                raise ValueError(f"Image key {key} must contain 'image'")
+            if features[key].type != "VISUAL":
+                raise ValueError(f"Image key {key} must map to a VISUAL feature")
+
+        # Ensure all image keys in features are in video_keys
+        feature_video_keys = [k for k in features.keys() if "image" in k.lower()]
+        if sorted(video_keys) != sorted(feature_video_keys):
+            raise ValueError(
+                "Video keys must include all image-related keys in features"
+            )
+
+        return self
+
+
+# Top-level model to validate the entire JSON
+class HuggingFaceModelValidator(BaseModel):
+    type: Literal["act"]
+    input_features: InputFeatures
+
+    class Config:
+        extra = "allow"
+
+
+class ACTSpawnConfig(BaseModel):
+    state_key: str
+    state_size: list[int]
+    video_keys: list[str]
+    video_size: list[int]
+    hf_model_config: HuggingFaceModelValidator
 
 
 class ACT(ActionModel):
     def __init__(
         self,
+        state_key: str,
+        state_size: list[int],
+        video_keys: list[str],
+        video_size: list[int],
         server_url: str = "http://localhost",
         server_port: int = 8080,
-        timeout: int = 10,
+        **kwargs,
     ):
         super().__init__(server_url, server_port)
-        self.required_input_keys: List[str] = ["images", "state"]
-        self.timeout = timeout
+        self.state_key = state_key
+        self.state_size = state_size
+        self.video_keys = video_keys
+        self.video_size = video_size
 
     def sample_actions(self, inputs: dict) -> np.ndarray:
         # Build the payload
         payload = {
-            "observation.state": inputs["state"],
+            self.state_key: inputs[self.state_key],
         }
-        for i in range(0, len(inputs["images"])):
-            payload[f"observation.images.{i}"] = inputs["images"][i]
+        for i in range(0, len(self.video_keys)):
+            payload[self.video_keys[i]] = inputs[self.video_keys[i]]
 
         # Double-encoded version (to send numpy arrays as JSON)
         encoded_payload = {"encoded": json_numpy.dumps(payload)}
 
         response = requests.post(
-            f"{self.server_url}:{self.server_port}/act",
+            f"{self.server_url}/act",
             json=encoded_payload,
-            timeout=self.timeout,
-        ).json()
+            timeout=10,
+        )
```

Comment:
add try except and log the error
make sure the HTTP Exception details are logged